t/lib/OpenQA/Test/Utils.pm

```diff
@@ -42,14 +44,37 @@ BEGIN {
 
 our (@EXPORT, @EXPORT_OK);
 @EXPORT_OK = (
+    qw(mock_service_ports),
     qw(redirect_output standard_worker create_user_for_workers),
     qw(create_webapi create_websocket_server create_scheduler create_live_view_handler),
-    qw(unresponsive_worker broken_worker rejective_worker wait_for_worker setup_share_dir setup_fullstack_temp_dir run_gru_job),
+    qw(unresponsive_worker broken_worker rejective_worker setup_share_dir setup_fullstack_temp_dir run_gru_job),
     qw(collect_coverage_of_gru_jobs stop_service start_worker unstable_worker fake_asset_server),
     qw(cache_minion_worker cache_worker_service shared_hash embed_server_for_testing),
     qw(run_cmd test_cmd)
 );
 
+# The function OpenQA::Utils::service_port method hardcodes ports in a
+# sequential range starting with OPENQA_BASE_PORT. This can cause problems
+# especially in repeated testing if any of the port in that range is already
+# occupied, so we inject random, free ports for the services here
+#
+# Potential point for
+# later improvement: In Mojo::IOLoop::Server::generate_port keep the sock
+# object on the port and reuse it in listen to prevent race condition
+#
+# Potentially this approach can also be used in production code.
```

Comment:
I suppose that would only work when we had a "master Perl script" to launch these services.

Reply:
I guess in the end the best approach is to just be less smart in our code and any code that spawns the services should be able to tell which port to use, be it a simple shell script spawing all processes in the background or something like k8s reading port variables from a cluster global config map.